shared/ui/textField/TextField.tsx

```diff
@@ -1,102 +1,123 @@
-import { forwardRef, useState } from 'react';
+import { ChangeEvent, ForwardedRef, ReactNode, forwardRef, useState } from 'react';
 
 import Close from '@/assets/icons/close.svg?url';
 import Eye from '@/assets/icons/eye-outline.svg?url';
 import SearchOutline from '@/assets/icons/searchOutline.svg?url';
 import { Typography } from '@/shared/ui/typography';
+import { clsx } from 'clsx';
 import Image from 'next/image';
 
 import s from 'shared/ui/textField/TextField.module.scss';
 
-// type TextFieldType = 'input' | 'textarea';
+type SharedProps = {
+  className?: string;
+  disabled?: boolean;
+  error?: string;
+  label?: string;
+  onBlur?: () => void;
+  onChange: (value: string) => void;
+  onFocus?: () => void;
+  placeholder?: string;
+  required?: boolean;
+  value: string;
+};
 
+// If you pass this type as a generic parameter, TS will prevent you from using props specific to union members.
+// To be able to pass these props, you can extract the desired union member using the Extract utility type.
+// https://www.typescriptlang.org/docs/handbook/utility-types.html#extracttype-union
 export type TextFieldProps =
-  // <T extends TextFieldType = 'input'>
+  | (SharedProps & {
+      as?: 'input';
+      inputType?: 'password' | 'search' | 'text';
+    })
+  | (SharedProps & { as: 'textarea' });
```

Comment:
This `TextFieldProps` syntax is called [discriminated union](https://scriptdev.ru/guide/033/).
>In order for type inference to distinguish one type included in a union set from another, it is necessary for each type to define a special field capable of identifying it uniquely.

In this case, it means that the value of the `as` prop will help Typescript to infer the other valid props, e.g. `inputType` for `as` is equal to `input`.